### PR TITLE
Added functions arrayFirstOrNull, arrayLastOrNull

### LINKED ISSUE #35238: ArrayFirstOrDefault function
For example, we have table like this
image

and i run query like this
```
select arrayFirst((i, v)->(v = 'a'), ids, values) as a_id,
       arrayFirst((i, v)->(v = 'c'), ids, values) as c_id
```
image

this result is not right for me, i need to write smth like 
```
select if(arrayCount((i, v)->(v = 'a'), ids, values) > 0, arrayFirst((i, v)->(v = 'a'), ids, values), null) as a_id,
       if(arrayCount((i, v)->(v = 'c'), ids, values) > 0, arrayFirst((i, v)->(v = 'c'), ids, values), null) as c_id
```

and will get right result
image

so, can we get function smth like `arrayFirstOrDefault `or `arrayFirstOrNull`?
```
select arrayFirstOrNull((i, v)->(v = 'a'), ids, values) as a_id,
       arrayFirstOrNull((i, v)->(v = 'c'), ids, values) as c_id
```

### PR DESCRIPTION
**Pull request by kitaisreal**
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added functions `arrayFirstOrNull`, `arrayLastOrNull`. Closes #35238.

### CODE REVIEW

File: src/Interpreters/RowRefs.cpp [hunk 1]
```diff
@@ -162,6 +162,12 @@ class SortedLookupVector : public SortedLookupVectorBase
             std::lock_guard<std::mutex> l(lock);
             if (!sorted.load(std::memory_order_relaxed))
             {
```
Comment (Algunenano) [inline src/Interpreters/RowRefs.cpp:164]: This seems unrelated to the changes. Have I broken something? 🤔

Comment (kitaisreal) [reply src/Interpreters/RowRefs.cpp]: No, just checked this file before.

File: src/Functions/array/arrayFirstLast.cpp [hunk 2]
```diff
@@ -67,16 +87,29 @@ struct ArrayFirstLastImpl
...
                 auto out = array.getData().cloneEmpty();
-                out->insertDefault();
-                return out->replicate(IColumn::Offsets(1, array.size()));
+                out->insertManyDefaults(array.size());
+
+                if constexpr (element_not_exists_strategy == ArrayFirstLastElementNotExistsStrategy::Null)
+                {
+                    auto col_null_map_to = ColumnUInt8::create(out->size(), true);
+                    return ColumnNullable::create(std::move(out), std::move(col_null_map_to));
+                }
```
Comment (vdimir) [inline src/Functions/array/arrayFirstLast.cpp:110]: Why we insert default before converting to nullable? Should we insert default or null here?

**Review by vdimir — APPROVED** [review]

File: src/Functions/array/arrayFirstLast.cpp (hunk 2 above)
Comment (kitaisreal) [reply src/Functions/array/arrayFirstLast.cpp:110]: Memory sanitizer will raise an error during access of uninitialized data when column serialization/deserialization will take place. It can be fixed to avoid performance penalty, but for this function it is okay.
